Makefile

```diff
@@ -187,12 +190,17 @@ endif
 # Steps to release: create release on github, git pull, git describe --tags to verify you're
 # on the release tag, make release, upload esp-link.tgz into the release files
 #VERSION ?= "esp-link custom version"
-DATE    := $(shell date '+%F %T')
+DATE    := $(shell date '+%F')
+TIME    := $(shell date '+%T') 
 BRANCH  ?= $(shell if git diff --quiet HEAD; then git describe --tags; \
                    else git symbolic-ref --short HEAD; fi)
 SHA     := $(shell if git diff --quiet HEAD; then git rev-parse --short HEAD | cut -d"/" -f 3; \
                    else echo "development"; fi)
-VERSION ?=esp-link $(BRANCH) - $(DATE) - $(SHA)
+VERSION ?=esp-link $(BRANCH) - $(DATE) - $(TIME) - $(SHA)
+
+BUILD_NUMBER_FILE = $(abspath ./build-number.txt)
+BUILD_NUMBER_SRC = $(abspath ./user/buildnum.c)
+BUILD_NUMBER := $(shell if ! test -f $(BUILD_NUMBER_FILE); then echo 1 > $(BUILD_NUMBER_FILE); fi; cat $(BUILD_NUMBER_FILE)) 
```

Comment:
These

- separate the date and time, because someone may need only the date, only the time, or both
- BUILD_NUMBER_FILE is a text file, which contains a single number - counting how many times the user has built the source code (issuing **make** commands)
- BUILD_NUMBER_SRC defines the C file that has to be rebuild whenever the BUILD_NUMBER changes - because this C file contains declarations for the textual representation of build date, build time and build number. Obviously the strings in C code should change every time one of the values (date, time, build number) changes.
- BUILD_NUMBER reads the number from the above specified text file (or creates a new file with value 1)

These 3 constants are used later in the Makefile.
Why having a build number may be better than only having date and time ? Because it is easier to remember one number than a whole timestamp - so when relating to a given build (which contains some new features), it is simpler to use a single sequential number.